File: library/builtins/traits/functions/closure.rs
use crate::*;
use wipple::*;

#[derive(TypeInfo, Clone)]
pub struct Closure {
    pub captured_env: Env,
    pub parameter_pattern: Pattern,
    pub parameter_name: String,
    pub return_value: Value,
}

impl Primitive for Closure {}

pub(crate) fn setup(env: &Env, stack: &Stack) -> Result<()> {
    // Closure == Function
    env.add_relation_between(stack, |closure: Closure| {
        Function::new(move |value, env, stack| {
            let value = value.evaluate(env, stack)?;

            let validated = (closure.parameter_pattern)(&value, env, stack)?;

            let value = validated
                .ok_or_else(|| error("Cannot use this value as input to this closure", stack))?;

            let inner_env = closure.captured_env.child();

            inner_env.set_variable(&closure.parameter_name, value.into_owned());

            let result = closure.return_value.evaluate(&inner_env, stack)?;

            Ok(result.into_owned())
        })
    })?;

    // Closure == Text
    env.add_text_relation::<Closure>("closure", stack)?;

    Ok(())
}

File: Library/Sources/Traits/Functions/Closure.swift
import Wipple

public struct Closure: Primitive {
    public var capturedEnv: Env
    public var parameter: (pattern: Pattern, name: String)
    public var returnValue: Value

    public init(
        capturing capturedEnv: Env,
        parameter: (pattern: Pattern, name: String),
        returning returnValue: Value
    ) {
        self.capturedEnv = capturedEnv
        self.parameter = parameter
        self.returnValue = returnValue
    }
}

internal func setupClosure(_ env: Env, _ stack: Stack) throws {
    // Closure == Function
    try env.addRelation(from: Closure.self, to: Function.self, stack) { closure in
        Function { value, env, stack in
            let value = try value.evaluate(env, stack)

            guard let matchedValue = try closure.parameter.pattern.match(value, env, stack) else {
                throw Exit.error("Cannot use this value as input to this closure", stack)
            }

            let closureEnv = closure.capturedEnv.child()
            try closureEnv.setVariable(closure.parameter.name, to: matchedValue, stack)

            return try closure.returnValue.evaluate(closureEnv, stack)
        }
    }

    // Closure == Text
    try env.addRelation(text: "closure", for: Closure.self, stack)
}
